### PR TITLE
Feature/saving predictions

### PR DESCRIPTION
**Pull request by asawczyn**

### CODE REVIEW

File: embeddings/evaluator/evaluator.py [hunk 1]
```diff
@@ -9,8 +9,8 @@
 
 
 class Evaluator(ABC, Generic[Input, Output]):
-    def __init__(self) -> None:
-        pass
+    def __init__(self, return_input_data: bool = True) -> None:
+        self.return_input_data = return_input_data
```
Comment (djaniak) [inline embeddings/evaluator/evaluator.py:13]: what is this parameter for? do you use it further in the code?

File: embeddings/pipeline/lightning_classification.py [hunk 2]
```diff
@@ -39,8 +41,19 @@ def __init__(
...
+        output_path = Path(output_path)
+        task_train_kwargs = (
+            task_train_kwargs if task_train_kwargs else self.DEFAULT_TASK_TRAIN_KWARGS
+        )
+        task_model_kwargs = (
+            task_model_kwargs if task_model_kwargs else self.DEFAULT_TASK_MODEL_KWARGS
+        )
+        tokenizer_name = tokenizer_name if tokenizer_name else embedding_name
+        datamodule_kwargs = datamodule_kwargs if datamodule_kwargs else {}
+        model_config_kwargs = model_config_kwargs if model_config_kwargs else {}
```
Comment (djaniak) [inline embeddings/pipeline/lightning_classification.py:53]: note that this is changed in #173 as these kwargs get updated now

File: embeddings/pipeline/lightning_sequence_labeling.py [hunk 3]
```diff
@@ -45,8 +47,20 @@ def __init__(
...
+        task_train_kwargs = (
+            task_train_kwargs if task_train_kwargs else self.DEFAULT_TASK_TRAIN_KWARGS
+        )
+        task_model_kwargs = (
+            task_model_kwargs if task_model_kwargs else self.DEFAULT_TASK_MODEL_KWARGS
+        )
+        tokenizer_name = tokenizer_name if tokenizer_name else embedding_name
+        datamodule_kwargs = (
+            datamodule_kwargs if datamodule_kwargs else self.DEFAULT_DATAMODULE_KWARGS
+        )
```
Comment (djaniak) [inline embeddings/pipeline/lightning_sequence_labeling.py:60]: note that this is changed in #173 as these kwargs get updated now

File: embeddings/utils/utils.py [hunk 4]
```diff
@@ -32,3 +37,35 @@ def import_from_string(dotted_path: str) -> Any:
...
         raise ImportError(msg)
+
+
+def build_output_path(root: T_path, embedding_name: T_path, dataset_name: T_path) -> Path:
+    """Builds output path using pattern {root}/{embedding_name}/{dataset_name}.
+    Every "/" in the embedding/dataset name is replaced with  "__".
+    E.g. "clarin-pl/nkjp-pos" -> "clarin-pl__nkjp-pos".
+
+    Be aware that passed paths is str (instead of Path) are not checked if they exist and if
+    they are dirs.
```
Comment (djaniak) [inline embeddings/utils/utils.py:48]:
```suggestion
    Be aware that if passed paths are str (instead of Path) they are not checked if they exist
    and if they are dirs.
```

File: pyproject.toml [hunk 5]
```diff
@@ -58,7 +58,7 @@ black_fix = "black ./"
 isort_fix = "isort . "
 pyflakes = "pyflakes embeddings"
 mypy = "mypy -p embeddings"
-coverage_test = "coverage run -m pytest"
+coverage_test = "coverage run -m pytest tests/test_lightning_classification_pipeline.py"
```
Comment (djaniak) [inline pyproject.toml:61]: what is this change for?

File: tests/test_aggregation_embeddings.py [hunk 6]
```diff
@@ -12,7 +12,7 @@
 from embeddings.embedding.static.word2vec import KGR10Word2VecConfig
 
 
-@pytest.fixture
+@pytest.fixture(scope="module")
```
Comment (djaniak) [inline tests/test_aggregation_embeddings.py:15]: All fixtures' scopes were changed to module level? What is the reason for this change? If I am correct, this fixture will be overwritten when defined in another file.

File: tests/test_output_path_builder.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+hub_repo_official = "hub-123"
+hub_repo_unofficial = "org/hub-123"
+
+
+@pytest.fixture(scope="module")
+def local_repo() -> Iterator[Path]:
+    with TemporaryDirectory() as temp_path:
+        local_repo = Path(temp_path).joinpath("local-123")
+        local_repo.mkdir()
+        yield local_repo
```
Comment (djaniak) [inline tests/test_output_path_builder.py:19]: if the scope range is module then it will be called only once; why do we use Iterator and yield the Path instead of simply returning it? I realize that the outcome will be the same

File: embeddings/evaluator/evaluator.py (hunk 1 above)
Comment (asawczyn) [reply embeddings/evaluator/evaluator.py]: It has been moved to MetricsEvaluator

File: embeddings/pipeline/lightning_classification.py (hunk 2 above)
Comment (asawczyn) [reply embeddings/pipeline/lightning_classification.py:53]: I have commented on changes in #173 and if I am right, this is a proper solution.

File: embeddings/evaluator/evaluator.py [hunk 8]
```diff
@@ -9,10 +8,6 @@
 
 
 class Evaluator(ABC, Generic[Input, Output]):
-    def __init__(self) -> None:
-        pass
-
-    @abc.abstractmethod
```
Comment (ktagowski) [inline embeddings/evaluator/evaluator.py:15]: What is purpose of that removal?

File: embeddings/pipeline/lightning_sequence_labeling.py (hunk 3 above)
Comment (asawczyn) [reply embeddings/pipeline/lightning_sequence_labeling.py:60]: ok

File: embeddings/evaluator/evaluator.py (hunk 8 above)
Comment (asawczyn) [reply embeddings/evaluator/evaluator.py:15]: fixes

**Review by ktagowski — DISMISSED** [review]

File: embeddings/utils/utils.py [hunk 9]
```diff
@@ -32,3 +37,35 @@ def import_from_string(dotted_path: str) -> Any:
...
+        embedding_name = Path(embedding_name).name
+    else:
+        assert isinstance(embedding_name, str)
+        embedding_name = embedding_name.replace("/", "__")
+
+    if os.path.isdir(dataset_name):
+        dataset_name = Path(dataset_name).name
+    else:
+        assert isinstance(dataset_name, str)
+        dataset_name = dataset_name.replace("/", "__")
```
Comment (djaniak) [inline embeddings/utils/utils.py:65]: could we pack this into a function? does same thing to different arguments

**Review by ktagowski — APPROVED** [review]

**Review by djaniak — APPROVED** [review]
